clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala

```diff
@@ -0,0 +1,244 @@
+        "left"
+      )
+      .select(
+        col("u.rangeID").as("unique_range"),
+        col("a.rangeID").as("active_range"),
+        functions.concat(col("u.min_key"), col("u.max_key")).as("min-max")
+      )
+
+    // group unique_ranges and active_ranges by min-max
+    // unique_ranges: Set[String] | active_ranges: Set[String] | min-max
+    // for each row, unique_ranges contains ranges with exact min-max, active_ranges contain ranges intersecting with min-max
+    val groupByMinMax = joinActiveByRange
+      .groupBy("min-max")
+      .agg(
+        collect_set("unique_range").alias("unique_ranges"),
+        collect_set("active_range").alias("active_ranges")
+      )
+
+    // add a column of all addresses contained only in unique ranges
+    // for every Row - [ collection of unique ranges | collection of active ranges] the column would contain a collection of addresses existing only in unique ranges
+    val addresses = groupByMinMax.withColumn(
+      "addresses",
+      left_anti_join_addresses(col("unique_ranges"), col("active_ranges"))
+    )
+
+    addresses
+      .select(explode(col("addresses")).as("addresses"))
+      .select(
+        col("addresses._1").as("key"),
+        col("addresses._2").as("address"),
+        col("addresses._3").as("relative"),
+        col("addresses._4").as("last_modified")
+      )
+  }
+
+  def getExpiredAddresses(
+      repo: String,
+      runID: String,
+      commitDFLocation: String,
+      spark: SparkSession,
+      conf: APIConfigurations
+  ): Dataset[Row] = {
+    val commitsDF = getCommitsDF(runID, commitDFLocation, spark)
+    val rangesDF = getRangesDFFromCommits(commitsDF, repo, conf)
+    getAddressesDFFromRanges(rangesDF, conf, repo)
+  }
+
+  def main(args: Array[String]) {
+    val spark = SparkSession.builder().getOrCreate()
+    if (args.length != 4) {
+      Console.err.println(
+        "Usage: ... <repo_name> <runID> s3://storageNamespace/prepared_commits_table s3://storageNamespace/output_destination_table"
+      )
+      System.exit(1)
+    }
+
+    val repo = args(0)
+    val runID = args(1)
+    val commitDFLocation = args(2)
+    val addressesDFLocation = args(3)
+
+    val hc = spark.sparkContext.hadoopConfiguration
+    val apiURL = hc.get(LAKEFS_CONF_API_URL_KEY)
+    val accessKey = hc.get(LAKEFS_CONF_API_ACCESS_KEY_KEY)
+    val secretKey = hc.get(LAKEFS_CONF_API_SECRET_KEY_KEY)
+    val expiredAddresses = getExpiredAddresses(repo,
+                                               runID,
+                                               commitDFLocation,
+                                               spark,
+                                               APIConfigurations(apiURL, accessKey, secretKey)
+                                              ).withColumn("run_id", lit(runID))
+    expiredAddresses.write.partitionBy("run_id").mode(SaveMode.Append).parquet(addressesDFLocation)
```

Comment:
I don't understand the `partitionBy` clause.  AFAICS column `"run_id"` is the constant value `runID`.  So what is this partition?  Why do we even need it?

Reply:
In order to write the data partitioned in The object store (i.e under `s3://path/run_id=1234/...`)

Reply:
So IIUC this is "just" a shortcut for naming the files as though they belonged to a partition, even though there's no partition here?  Not enough of a Spark person to know whether users will expect this -- so let's try and see!

Reply:
Yes the parquet table in `addressesDFLocation` is partitioned by `run_id` the results of this run should be under the corresponding partition.